File: python/src/memosift/core/resolution_tracker.py
# Resolution tracker — audit-only detection of question→decision arcs in conversations.
#
# IMPORTANT: This module is READ-ONLY — it detects patterns and reports them
# but does NOT modify shields, scores, or any compression behavior. It exists
# to gather data on whether semantic detection would improve compression quality,
# without risking regressions from false positives.
from __future__ import annotations

import re
from dataclasses import dataclass, field
from dataclasses import replace as dc_replace
from typing import TYPE_CHECKING

from memosift.core.types import CompressionPolicy

if TYPE_CHECKING:
    from memosift.core.types import ClassifiedMessage

# Question detection patterns.
_QUESTION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\?\s*$", re.MULTILINE),  # Ends with ?
    re.compile(r"\bshould (?:we|I)\b", re.IGNORECASE),
    re.compile(r"\bwhich (?:one|approach|option)\b", re.IGNORECASE),
    re.compile(r"\bhow (?:to|should|do)\b", re.IGNORECASE),
    re.compile(r"\bwhat (?:about|if)\b", re.IGNORECASE),
]

# Deliberation patterns (comparison, alternatives).
_DELIBERATION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\balternatively\b", re.IGNORECASE),
    re.compile(r"\bon (?:one|the other) hand\b", re.IGNORECASE),
    re.compile(r"\bpros and cons\b", re.IGNORECASE),
    re.compile(r"\boption [A-Z]\b", re.IGNORECASE),
    re.compile(r"\bcould (?:also |either )?use\b", re.IGNORECASE),
    re.compile(r"\bvs\.?\b", re.IGNORECASE),
    re.compile(r"\bcompare\b", re.IGNORECASE),
    re.compile(r"\btradeoff\b", re.IGNORECASE),
]

# Decision/resolution patterns (reuses anchor_extractor's markers).
_RESOLUTION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\bI'll use\b", re.IGNORECASE),
    re.compile(r"\bLet's go with\b", re.IGNORECASE),
    re.compile(r"\bchoosing\b.{1,60}\bbecause\b", re.IGNORECASE),
    re.compile(r"\bdecided to\b", re.IGNORECASE),
    re.compile(r"\bI'll go with\b", re.IGNORECASE),
    re.compile(r"\bwe(?:'ll| will) use\b", re.IGNORECASE),
    re.compile(r"\bgoing with\b", re.IGNORECASE),
    re.compile(r"\bthe (?:best|right) (?:choice|approach|option) is\b", re.IGNORECASE),
]

# Supersession patterns (later message corrects/updates earlier).
_SUPERSESSION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\bactually\b", re.IGNORECASE),
    re.compile(r"\bturns out\b", re.IGNORECASE),
    re.compile(r"\bcorrection\b", re.IGNORECASE),
    re.compile(r"\bnot .{1,30} but\b", re.IGNORECASE),
    re.compile(r"\bupdated?\b.*\bnow\b", re.IGNORECASE),
    re.compile(r"\ball \d+ tests? pass", re.IGNORECASE),  # Status update pattern
]

# Stop words for keyword extraction.
_STOP_WORDS: frozenset[str] = frozenset(
    {
        "a",
        "an",
        "the",
        "and",
        "or",
        "but",
        "in",
        "on",
        "at",
        "to",
        "for",
        "of",
        "with",
        "by",
        "from",
        "is",
        "are",
        "was",
        "were",
        "be",
        "been",
        "have",
        "has",
        "had",
        "do",
        "does",
        "did",
        "will",
        "would",
        "could",
        "should",
        "may",
        "might",
        "can",
        "not",
        "no",
        "if",
        "then",
        "else",
        "this",
        "that",
        "these",
        "those",
        "it",
        "its",
        "we",
        "you",
        "they",
        "use",
        "using",
        "used",
        "going",
        "think",
        "want",
        "need",
        "like",
    }
)


@dataclass(frozen=True)
class ResolutionArc:
    """A detected question→deliberation→decision arc. Audit-only — no compression effect."""

    question_index: int
    deliberation_indices: tuple[int, ...]
    resolution_index: int | None
    resolved: bool
    topic_keywords: frozenset[str]


@dataclass(frozen=True)
class SupersessionSignal:
    """A detected supersession (later message corrects/updates earlier). Audit-only."""

    superseded_index: int
    superseding_index: int
    reason: str  # "correction", "status_update", "refinement"
    shared_entities: frozenset[str]


@dataclass
class ResolutionReport:
    """Summary of detected arcs and supersessions. Logged to CompressionReport."""

    arcs: list[ResolutionArc] = field(default_factory=list)
    supersessions: list[SupersessionSignal] = field(default_factory=list)

    @property
    def resolved_count(self) -> int:
        return sum(1 for a in self.arcs if a.resolved)

    @property
    def unresolved_count(self) -> int:
        return sum(1 for a in self.arcs if not a.resolved)

    def to_dict(self) -> dict:
        """Serialize for CompressionReport."""
        return {
            "arcs_detected": len(self.arcs),
            "arcs_resolved": self.resolved_count,
            "arcs_unresolved": self.unresolved_count,
            "supersessions_detected": len(self.supersessions),
            "supersession_reasons": dict(_count_by(s.reason for s in self.supersessions)),
        }


def detect_resolution_arcs(
    segments: list[ClassifiedMessage],
) -> ResolutionReport:
    """Detect question→deliberation→decision arcs in classified messages.

    This is AUDIT-ONLY — it returns a report but does NOT modify any
    segment's shield, relevance score, or compression behavior.

    Args:
        segments: Classified messages from L1.

    Returns:
        ResolutionReport with detected arcs and supersessions.
    """
    report = ResolutionReport()

    # Step 1: Identify message roles (question, deliberation, resolution).
    questions: list[int] = []
    deliberations: list[int] = []
    resolutions: list[int] = []

    for i, seg in enumerate(segments):
        content = seg.content
        if not content or len(content) < 10:
            continue

        role = seg.message.role

        # Questions come from user messages.
        if role == "user" and _matches_any(content, _QUESTION_PATTERNS):
            questions.append(i)

        # Deliberation and resolution come from assistant messages.
        if role == "assistant":
            if _matches_any(content, _RESOLUTION_PATTERNS):
                resolutions.append(i)
            elif _matches_any(content, _DELIBERATION_PATTERNS):
                deliberations.append(i)

    # Step 2: Link questions to resolutions via keyword overlap.
    for q_idx in questions:
        q_keywords = _extract_keywords(segments[q_idx].content)
        if len(q_keywords) < 2:
            continue

        # Find deliberations between this question and the next question.
        next_q = min((qi for qi in questions if qi > q_idx), default=len(segments))
        arc_deliberations = [d for d in deliberations if q_idx < d < next_q]

        # Find the first resolution after this question with keyword overlap.
        arc_resolution: int | None = None
        for r_idx in resolutions:
            if r_idx <= q_idx:
                continue
            r_keywords = _extract_keywords(segments[r_idx].content)
            overlap = q_keywords & r_keywords
            if len(overlap) >= 2:  # Require at least 2 shared non-stop-word tokens.
                arc_resolution = r_idx
                break

        resolved = arc_resolution is not None
        topic = q_keywords
        if arc_resolution is not None:
            topic = topic | _extract_keywords(segments[arc_resolution].content)

        report.arcs.append(
            ResolutionArc(
                question_index=q_idx,
                deliberation_indices=tuple(arc_deliberations),
                resolution_index=arc_resolution,
                resolved=resolved,
                topic_keywords=frozenset(topic),
            )
        )

    # Step 3: Detect supersessions (corrections, status updates).
    for i, seg in enumerate(segments):
        if seg.message.role != "assistant" or len(seg.content) < 20:
            continue
        if not _matches_any(seg.content, _SUPERSESSION_PATTERNS):
            continue

        # Look for an earlier message with shared entities.
        i_entities = _extract_entities(seg.content)
        if not i_entities:
            continue

        for j in range(max(0, i - 20), i):
            if segments[j].message.role not in ("assistant", "tool"):
                continue
            j_entities = _extract_entities(segments[j].content)
            shared = i_entities & j_entities
            if len(shared) >= 1:
                reason = _classify_supersession(seg.content)
                report.supersessions.append(
                    SupersessionSignal(
                        superseded_index=j,
                        superseding_index=i,
                        reason=reason,
                        shared_entities=frozenset(shared),
                    )
                )
                break  # One supersession per message.

    return report


def apply_resolution_compression(
    segments: list[ClassifiedMessage],
    report: ResolutionReport,
) -> list[ClassifiedMessage]:
    """Apply compression policy changes based on detected resolution arcs.

    For resolved arcs: deliberation messages get AGGRESSIVE policy, the
    resolution itself gets its shield elevated. For superseded messages:
    the superseded message gets AGGRESSIVE policy.

    This function is opt-in, gated by ``config.enable_resolution_compression``.

    Args:
        segments: Classified messages from the pipeline.
        report: Resolution report from ``detect_resolution_arcs()``.

    Returns:
        Segments with updated compression policies.
    """
    result = list(segments)

    # Collect indices to mark as AGGRESSIVE.
    aggressive_indices: set[int] = set()

    # Resolved arcs: deliberation messages become AGGRESSIVE.
    for arc in report.arcs:
        if arc.resolved:
            for d_idx in arc.deliberation_indices:
                if d_idx < len(result):
                    aggressive_indices.add(d_idx)

    # Superseded messages become AGGRESSIVE.
    for sup in report.supersessions:
        if sup.superseded_index < len(result):
            aggressive_indices.add(sup.superseded_index)

    # Apply changes.
    for idx in aggressive_indices:
        seg = result[idx]
        if seg.policy not in {CompressionPolicy.PRESERVE, CompressionPolicy.LIGHT}:
            result[idx] = dc_replace(seg, policy=CompressionPolicy.AGGRESSIVE)

    return result


def _matches_any(text: str, patterns: list[re.Pattern[str]]) -> bool:
    """Return True if any pattern matches the text."""
    return any(p.search(text) for p in patterns)


def _extract_keywords(text: str) -> frozenset[str]:
    """Extract non-stop-word tokens from text."""
    tokens = set(re.findall(r"\b[a-zA-Z]\w{2,}\b", text.lower()))
    return frozenset(tokens - _STOP_WORDS)


def _extract_entities(text: str) -> frozenset[str]:
    """Extract likely entity names: file paths, camelCase, function names."""
    entities: set[str] = set()
    # File paths.
    for m in re.finditer(r"[\w./\\]+\.\w{1,8}", text):
        entities.add(m.group(0).lower())
    # camelCase / PascalCase.
    for m in re.finditer(r"\b[a-z]+(?:[A-Z][a-z]+)+\b", text):
        entities.add(m.group(0).lower())
    # Function-like patterns.
    for m in re.finditer(r"\b(\w+)\s*\(", text):
        name = m.group(1).lower()
        if name not in _STOP_WORDS and len(name) > 2:
            entities.add(name)
    return frozenset(entities)


def _classify_supersession(text: str) -> str:
    """Classify the type of supersession."""
    lower = text.lower()
    if any(w in lower for w in ("actually", "turns out", "correction", "not", "but")):
        return "correction"
    if any(w in lower for w in ("all", "pass", "now", "updated")):
        return "status_update"
    return "refinement"


def _count_by(items):
    """Count items by value."""
    counts: dict[str, int] = {}
    for item in items:
        counts[item] = counts.get(item, 0) + 1
    return counts

File: typescript/src/core/resolution-tracker.ts
// Resolution tracker — audit-only detection of question→decision arcs in conversations.
//
// IMPORTANT: This module is READ-ONLY — it detects patterns and reports them
// but does NOT modify shields, scores, or any compression behavior. It exists
// to gather data on whether semantic detection would improve compression quality,
// without risking regressions from false positives.

import { CompressionPolicy, type ClassifiedMessage } from "./types.js";

// Question detection patterns.
const QUESTION_PATTERNS = [
  /\?\s*$/m,
  /\bshould (?:we|I)\b/i,
  /\bwhich (?:one|approach|option)\b/i,
  /\bhow (?:to|should|do)\b/i,
  /\bwhat (?:about|if)\b/i,
];

// Deliberation patterns.
const DELIBERATION_PATTERNS = [
  /\balternatively\b/i,
  /\bon (?:one|the other) hand\b/i,
  /\bpros and cons\b/i,
  /\boption [A-Z]\b/i,
  /\bcould (?:also |either )?use\b/i,
  /\bvs\.?\b/i,
  /\bcompare\b/i,
  /\btradeoff\b/i,
];

// Resolution patterns (mirrors anchor_extractor decision markers).
const RESOLUTION_PATTERNS = [
  /\bI'll use\b/i,
  /\bLet's go with\b/i,
  /\bchoosing\b.{1,60}\bbecause\b/i,
  /\bdecided to\b/i,
  /\bI'll go with\b/i,
  /\bwe(?:'ll| will) use\b/i,
  /\bgoing with\b/i,
  /\bthe (?:best|right) (?:choice|approach|option) is\b/i,
];

// Supersession patterns.
const SUPERSESSION_PATTERNS = [
  /\bactually\b/i,
  /\bturns out\b/i,
  /\bcorrection\b/i,
  /\bnot .{1,30} but\b/i,
  /\bupdated?\b.*\bnow\b/i,
  /\ball \d+ tests? pass/i,
];

const STOP_WORDS = new Set([
  "a",
  "an",
  "the",
  "and",
  "or",
  "but",
  "in",
  "on",
  "at",
  "to",
  "for",
  "of",
  "with",
  "by",
  "from",
  "is",
  "are",
  "was",
  "were",
  "be",
  "been",
  "have",
  "has",
  "had",
  "do",
  "does",
  "did",
  "will",
  "would",
  "could",
  "should",
  "may",
  "might",
  "can",
  "not",
  "no",
  "if",
  "then",
  "else",
  "this",
  "that",
  "these",
  "those",
  "it",
  "its",
  "we",
  "you",
  "they",
  "use",
  "using",
  "used",
  "going",
  "think",
  "want",
  "need",
  "like",
]);

export interface ResolutionArc {
  readonly questionIndex: number;
  readonly deliberationIndices: readonly number[];
  readonly resolutionIndex: number | null;
  readonly resolved: boolean;
  readonly topicKeywords: ReadonlySet<string>;
}

export interface SupersessionSignal {
  readonly supersededIndex: number;
  readonly supersedingIndex: number;
  readonly reason: string;
  readonly sharedEntities: ReadonlySet<string>;
}

export interface ResolutionReport {
  readonly arcs: readonly ResolutionArc[];
  readonly supersessions: readonly SupersessionSignal[];
}

export interface ResolutionSignals {
  arcsDetected: number;
  arcsResolved: number;
  arcsUnresolved: number;
  supersessionsDetected: number;
  supersessionReasons: Record<string, number>;
}

/** Convert a ResolutionReport to a plain signals object for CompressionReport. */
export function toSignals(report: ResolutionReport): ResolutionSignals {
  const reasons: Record<string, number> = {};
  for (const s of report.supersessions) {
    reasons[s.reason] = (reasons[s.reason] ?? 0) + 1;
  }
  return {
    arcsDetected: report.arcs.length,
    arcsResolved: report.arcs.filter((a) => a.resolved).length,
    arcsUnresolved: report.arcs.filter((a) => !a.resolved).length,
    supersessionsDetected: report.supersessions.length,
    supersessionReasons: reasons,
  };
}

/**
 * Detect question→deliberation→decision arcs in classified messages.
 *
 * This is AUDIT-ONLY — it returns a report but does NOT modify any
 * segment's shield, relevance score, or compression behavior.
 */
export function detectResolutionArcs(segments: ClassifiedMessage[]): ResolutionReport {
  const questions: number[] = [];
  const deliberations: number[] = [];
  const resolutions: number[] = [];

  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i]!;
    const content = seg.message.content;
    if (!content || content.length < 10) continue;
    const role = seg.message.role;

    if (role === "user" && matchesAny(content, QUESTION_PATTERNS)) {
      questions.push(i);
    }
    if (role === "assistant") {
      if (matchesAny(content, RESOLUTION_PATTERNS)) {
        resolutions.push(i);
      } else if (matchesAny(content, DELIBERATION_PATTERNS)) {
        deliberations.push(i);
      }
    }
  }

  // Link questions to resolutions via keyword overlap.
  const arcs: ResolutionArc[] = [];
  for (const qIdx of questions) {
    const qKeywords = extractKeywords(segments[qIdx]!.message.content);
    if (qKeywords.size < 2) continue;

    const nextQ = Math.min(...questions.filter((qi) => qi > qIdx), segments.length);
    const arcDelibs = deliberations.filter((d) => d > qIdx && d < nextQ);

    let arcResolution: number | null = null;
    for (const rIdx of resolutions) {
      if (rIdx <= qIdx) continue;
      const rKeywords = extractKeywords(segments[rIdx]!.message.content);
      const overlap = new Set([...qKeywords].filter((k) => rKeywords.has(k)));
      if (overlap.size >= 2) {
        arcResolution = rIdx;
        break;
      }
    }

    const topic = new Set(qKeywords);
    if (arcResolution !== null) {
      for (const k of extractKeywords(segments[arcResolution]!.message.content)) {
        topic.add(k);
      }
    }

    arcs.push({
      questionIndex: qIdx,
      deliberationIndices: arcDelibs,
      resolutionIndex: arcResolution,
      resolved: arcResolution !== null,
      topicKeywords: topic,
    });
  }

  // Detect supersessions.
  const supersessions: SupersessionSignal[] = [];
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i]!;
    if (seg.message.role !== "assistant" || seg.message.content.length < 20) continue;
    if (!matchesAny(seg.message.content, SUPERSESSION_PATTERNS)) continue;

    const iEntities = extractEntities(seg.message.content);
    if (iEntities.size === 0) continue;

    for (let j = Math.max(0, i - 20); j < i; j++) {
      const prev = segments[j]!;
      if (prev.message.role !== "assistant" && prev.message.role !== "tool") continue;
      const jEntities = extractEntities(prev.message.content);
      const shared = new Set([...iEntities].filter((e) => jEntities.has(e)));
      if (shared.size >= 1) {
        supersessions.push({
          supersededIndex: j,
          supersedingIndex: i,
          reason: classifySupersession(seg.message.content),
          sharedEntities: shared,
        });
        break;
      }
    }
  }

  return { arcs, supersessions };
}

function matchesAny(text: string, patterns: RegExp[]): boolean {
  return patterns.some((p) => p.test(text));
}

function extractKeywords(text: string): Set<string> {
  const tokens = new Set(text.toLowerCase().match(/\b[a-zA-Z]\w{2,}\b/g) ?? []);
  for (const sw of STOP_WORDS) tokens.delete(sw);
  return tokens;
}

function extractEntities(text: string): Set<string> {
  const entities = new Set<string>();
  for (const m of text.matchAll(/[\w./\\]+\.\w{1,8}/g)) entities.add(m[0].toLowerCase());
  for (const m of text.matchAll(/\b[a-z]+(?:[A-Z][a-z]+)+\b/g)) entities.add(m[0].toLowerCase());
  for (const m of text.matchAll(/\b(\w+)\s*\(/g)) {
    const name = m[1]!.toLowerCase();
    if (!STOP_WORDS.has(name) && name.length > 2) entities.add(name);
  }
  return entities;
}

function classifySupersession(text: string): string {
  const lower = text.toLowerCase();
  if (["actually", "turns out", "correction", "not", "but"].some((w) => lower.includes(w))) {
    return "correction";
  }
  if (["all", "pass", "now", "updated"].some((w) => lower.includes(w))) {
    return "status_update";
  }
  return "refinement";
}

/** Apply compression policy changes based on detected resolution arcs.
 * For resolved arcs: deliberation messages get AGGRESSIVE policy.
 * For superseded messages: get AGGRESSIVE policy.
 * Opt-in, gated by config.enableResolutionCompression. */
export function applyResolutionCompression(
  segments: ClassifiedMessage[],
  report: ResolutionReport,
): ClassifiedMessage[] {
  const result = [...segments];
  const aggressiveIndices = new Set<number>();

  for (const arc of report.arcs) {
    if (arc.resolved) {
      for (const dIdx of arc.deliberationIndices) {
        if (dIdx < result.length) aggressiveIndices.add(dIdx);
      }
    }
  }
  for (const sup of report.supersessions) {
    if (sup.supersededIndex < result.length) {
      aggressiveIndices.add(sup.supersededIndex);
    }
  }

  for (const idx of aggressiveIndices) {
    const seg = result[idx]!;
    if (seg.policy !== CompressionPolicy.PRESERVE && seg.policy !== CompressionPolicy.LIGHT) {
      result[idx] = { ...seg, policy: CompressionPolicy.AGGRESSIVE };
    }
  }
  return result;
}
